src/time-slot-bar/time-slot-bar.tsx

```diff
@@ -0,0 +1,307 @@
+            }
+        };
+    }, []);
+
+    // =============================================================================
+    // HANDLERS
+    // =============================================================================
+
+    const handleScroll = (direction: "left" | "right") => {
+        if (barRef.current) {
+            barRef.current.scrollBy({
+                left:
+                    direction === "left" ? -SCROLL_INCREMENT : SCROLL_INCREMENT,
+                behavior: "smooth",
+            });
+        }
+    };
+
+    const handleResize = () => {
+        if (barRef.current) {
+            setClientWidth(barRef.current.clientWidth);
+        }
+    };
+
+    // ===========================================================================
+    // HELPER FUNCTIONS
+    // ===========================================================================
+
+    // Helper function to calculate the difference in minutes between two times
+    const getTimeDiffInMinutes = (start: string, end: string) => {
+        const startTime = dayjs(start, "HH:mm");
+        const endTime = dayjs(end, "HH:mm");
+        return endTime.diff(startTime, "minute");
+    };
+
+    // Helper function to convert minutes to a time string in HH:mm format
+    const minutesToTimeString = (minutes: number) => {
+        const hours = Math.floor(minutes / 60);
+        const remainingMinutes = minutes % 60;
+        return `${String(hours).padStart(2, "0")}:${String(
+            remainingMinutes
+        ).padStart(2, "0")}`;
+    };
+
+    // Helper function to format HH:mm to am/pm display
+    const formatHourlyDisplay = (timeString: string) => {
+        const parsedTime = dayjs(timeString, "HH:mm");
+        return parsedTime.format("ha");
+    };
+
+    // Helper compute width in px base on the time range in respect the cell width
+    const calculateWidth = (start: string, end: string) => {
+        /**
+         * Each CELL_WIDTH is 30min interval
+         * Assuming 15 minutes corresponds to a fixed width of 20 pixels (1/2 of CELL_WIDTH)
+         */
+        return (getTimeDiffInMinutes(start, end) / 15) * (CELL_WIDTH / 2);
+    };
+
+    // Speical Condition: Show only Ellipsis if the slotWidth is smaller or equal to a 15min interval
+    const showFullEllipsis = (slotWidth: number) => {
+        return slotWidth <= CELL_WIDTH / 2;
+    };
+
+    const getDataTestId = (subStr: string) => {
+        if (!testId) return undefined;
+        return `${testId}-${subStr}`;
+    };
+
+    // =============================================================================
+    // RENDER FUNCTIONS
+    // =============================================================================
+
+    // Render time markers
+    const renderTimeMarkers = () => {
+        const timeMarkers = [];
+        let currentTime = startTime;
+        while (getTimeDiffInMinutes(currentTime, endTime) >= CELL_DURATION) {
```

Comment:
I thought one other slightly simpler way we can do this is to do this
```suggestion
        const startTimeFormatted = dayjs(startTime, "HH:mm");
        const endTimeFormatted = dayjs(endTime, "HH:mm");

        for (
            let i = startTimeFormatted;
            i <= endTimeFormatted;
            i = i.add(CELL_DURATION, "minute")
        ) {
            const isHour = i.startOf("hour");
            // Render...
        }
```